File: Deepspace_Ruby/P2/Damage.rb
# BOLÍVAR PELÁEZ, CLARA
# RUIZ GÓMEZ, SOLEDAD
# 2º DGIIM - PDOO - Practice 2 DeepSpace

# Damage
# It represents the damage that an enemy starship makes to a
# spacial station when it loses a combat

module DeepSpace

    class Damage

        attr_reader :nShields, :nWeapons, :weapons
        
        # nWeapons constructor -> numeric damage
        def newNumericWeapons (w, s)
            @nShields = s 
            @nWeapons = w
            @weapons=Array.new()
            @weapons=nil
        end

        # WeaponType constructor -> specific damage 
        def newSpecificWeapons(w,s)
            @weapons=Array.new(w)
            @nShields= s
            @nWeapons= -1
            
        end

        # Copy constructor
        def newCopy(d)
            return d.clone
        end


        # builds a new DamageToUI object from *this
        def getUIversion (d)
            DamageToUI.new(self)
        end
        
        # ----
        def arrayContainsType (w, t)
            
            pos=0;

            w.each do |aux|
                if(aux.type == t)
                    return pos
                else
                    pos+=1
                end
            end

            return -1

        end

        # returns a an adjusted version of *this. It takes into consideration
        # the parameters w ans sb in order to modify the output value so it does not
        # imply losing weapons or shields that are nos specified in w or sb
        def adjust (w, s)
            shields = [@nShields, s.lenght].min 

            # If it is numeric damage
            if (@nWeapons != -1)
                output = Damage.new([@nWeapons, w.lenght].min, shields)
                return output
            else
                aux = []
                aux2 = w.clone

                weapons.each do |element|
                    i = arrayContainsType(aux2, element)
                    if (i != -1)
                        aux2.delete_at(i)
                        aux << element
                    end
                end
                output = Damage.new(aux, shields)
                return output
            end
        end


        # If *this has w.getType() in the array weapons, it deletes that element
        # of the array. In other case, it decrements nWeapons in one unit
        def discardWeapon (w)
            if (@weapons.delete(w) == nil)
                if (@nWeapons>0)
                    @nWeapons -= 1
                end
            end
        end

        # it decrements nShields in one unit
        def discardShieldBooster
            if (@nShields>0)
                @nShields -= 1
            end
        end

        # returns true if *this does not imply any accessory loss
        def hasNoEffect
            return ((@nWeapons==0) && (@nShields==0) && (@weapons.empty?))
        end

        #toString
        def to_s
            output = "Damage [ nShields #{@nShields} ; " +
                     "nWeapons #{@nWeapons} ; " +
                     "weapons  #{@weapons.to_s} ]"
        end

    end
end


        

File: Deepspace_Java/P3/src/deepspace/Damage.java
/**
 * @author BOLÍVAR PELÁEZ, CLARA
 * @author RUIZ GÓMEZ, SOLEDAD
 * 2º DGIIM - PDOO - Practice 1 DeepSpace
 */
package deepspace;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * @title Damage
 * @brief It represents the damage that an enemy starship makes to a
 * spacial station when it loses a combat
 */
public class Damage {
    
    private int nShields;
    private int nWeapons;
    private ArrayList<WeaponType> weapons;
    
    
    /**
    * @brief nWeapons constructor -> numeric damage
    */
    Damage (int w, int s) {
        this.nWeapons = w;
        this.nShields = s;
        this.weapons = new ArrayList<>(); 
    }
    
    /**
    * @brief WeaponType constructor -> specific damage
    */
    Damage (ArrayList<WeaponType> w, int s) {
        this.weapons = new ArrayList<WeaponType>(w);
        this.nShields = s;
        this.nWeapons = -1;        
    }
    
    /**
    * @brief Copy constructor
    * @param d another Damage instance
    */
    Damage (Damage d) {
        this.nWeapons = d.nWeapons;
        this.nShields = d.nShields;
        this.weapons = new ArrayList<>(d.weapons);         
    }
    
    /**
    * @brief builds a new DamageToUI object from *this
    * @return DamageToUI
    */
    DamageToUI getUIversion () {
        return new DamageToUI (this);
    }
    
    /**
    * @brief builds a new DamageToUI object from *this
    * @return DamageToUI
    */
    private int arrayContainsType (ArrayList <Weapon> w, WeaponType t) {
        
        Iterator<Weapon> it =w.iterator();
        int pos=0;
        
        while(it.hasNext()){
            Weapon aux = it.next();
            
            if(aux.getType() == t)
                return pos;
            else
                pos++;
        }
        
        return -1;
            
            
    }
    
    /**
    * @brief returns a an adjusted version of *this. It takes into consideration
    * the parameters w ans sb in order to modify the output value so it does not
    * imply losing weapons or shields that are nos specified in w or sb
    * @param w ArrayList of weapons
    * @param sb ArrayList of shield boosters
    * @return modifies version of *this
    */
     Damage adjust (ArrayList<Weapon> w, ArrayList<ShieldBooster> sb) {
        
        int shields = Integer.min(nShields, sb.size());

        // If it is numeric damage
        if (nWeapons != -1) {
            Damage output = new Damage (Integer.min (nWeapons, w.size()), shields);
            return output;
        // If it is specific damage
        } else {  
            ArrayList<WeaponType> aux = new ArrayList <> ();
            ArrayList<Weapon> aux2 = new ArrayList <> (w);
            
            for (WeaponType element: weapons) { // going through *this
                // checking if each type is contained in the param
                int i = arrayContainsType (aux2, element);
                if (i != -1) {
                    aux2.remove(i);
                    aux.add(element);
                }
            }
            
            Damage output = new Damage (aux, shields);
            return output;
        }
        
    }
    
    /**
    * @brief If *this has w.getType() in the array weapons, it deletes that element
    * of the array. In other case, it decrements nWeapons in one unit
    * @param w weapon which type is to be deleted
    * @return void
    */
    void discardWeapon (Weapon w) {
        if (!weapons.remove(w.getType()))
            if (nWeapons > 0)
                nWeapons--;     
    }
    
    /**
    * @brief it decrements nShields in one unit
    * @return void
    */
    void discardShieldBooster () {
        if (nShields > 0)
            nShields--;
    }
    
    /**
    * @brief returns true if *this does not imply any accessory loss
    * @return true if *this damage has no effect
    */
    boolean hasNoEffect () {
        return ((nWeapons==0) && (nShields==0) && (weapons.isEmpty()));
    }
    
    /**
    * @brief builds a new DamageToUI object from *this
    * @return DamageToUI
    */
    int getNShields () {
       return nShields;
    }
    
    /**
    * @brief builds a new DamageToUI object from *this
    * @return DamageToUI
    */
    int getNWeapons () {
      return nWeapons;        
    }
    
    /**
    * @brief builds a new DamageToUI object from *this
    * @return DamageToUI
    */
    ArrayList<WeaponType> getWeapons() {
        ArrayList<WeaponType> output = new ArrayList<>(weapons);
        return output;
    }
    
    
    /**
    * @brief toString
    * @return String with info about the instance
    */
    public String toString () {
        
        String output = "Damage [ nShields " + nShields 
                        + " ; nWeapons " + nWeapons
                        + " ; weapons " + weapons.toString() + " ]";
        
        return output;
    }
  
}
